File: samples/TapdaqExamples/Mediation/ViewController.m
//
//  ViewController.m
//  Mediation
//
//  Created by Nick Reffitt on 15/05/2017.
//  Copyright © 2017 Tapdaq. All rights reserved.
//

#import "ViewController.h"

NSString *AdTypeNameFromTDAdTypes(TDAdTypes types) {
    switch (types) {
        case TDAdTypeInterstitial:
            return @"interstitial";
        case TDAdTypeVideo:
            return @"video interstitial";
        case TDAdTypeRewardedVideo:
            return @"rewarded video";
        case TDAdTypeBanner:
            return @"banner";
        default:
            return @"unknown";
    }
}

static NSString *const kRewardNameKey = @"MyRewardName";
static NSString *const kRewardAmountKey = @"MyRewardAmount";
static NSString *const kRewardPayloadKey = @"MyRewardPayload";


@interface ViewController () <TapdaqDelegate, TDAdRequestDelegate, TDDisplayableAdRequestDelegate, TDClickableAdRequestDelegate, TDBannerAdRequestDelegate, TDRewardedVideoAdRequestDelegate>

@property (strong, nonatomic) TDMediationAdRequest *interstitialAdRequest;
@property (strong, nonatomic) TDMediationAdRequest *videoAdRequest;
@property (strong, nonatomic) TDMediationAdRequest *rewardedAdRequest;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [[Tapdaq sharedSession] setDelegate:self];
    [self updateRewardUI];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self updateUI];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)updateUI {
    self.loadInterstitialBtn.enabled = Tapdaq.sharedSession.isConfigLoaded;
    self.loadVideoBtn.enabled = Tapdaq.sharedSession.isConfigLoaded;
    self.loadRewardedBtn.enabled = Tapdaq.sharedSession.isConfigLoaded;
    self.loadBannerBtn.enabled = Tapdaq.sharedSession.isConfigLoaded;
    
    self.showInterstitialBtn.enabled = self.interstitialAdRequest.isReady;
    self.showVideoBtn.enabled = self.videoAdRequest.isReady;
    self.showRewardedBtn.enabled = self.rewardedAdRequest.isReady;
    
    self.showBannerBtn.enabled = [Tapdaq.sharedSession isBannerReady];
}

#pragma mark - Target action 

- (IBAction)loadBanner:(id)sender {
    [[Tapdaq sharedSession] loadBanner:TDMBannerStandard];
}

- (IBAction)showBanner:(id)sender
{
    NSInteger tagInt = [self.showBannerBtn tag];
    
    if (tagInt == 1) {
        
        [self.adBanner removeFromSuperview];
        
        [self.showBannerBtn setTag:0];
        [self.showBannerBtn setTitle:@"Show" forState:UIControlStateNormal];
        [self.showBannerBtn setEnabled:NO];
        
    } else {
        
        if ([[Tapdaq sharedSession] isBannerReady]) {
            
            self.adBanner = [[Tapdaq sharedSession] getBanner];
            [self.bannerView addSubview:self.adBanner];
            self.adBanner.center = self.bannerView.center;
            
            [self.showBannerBtn setTag:1];
            [self.showBannerBtn setTitle:@"Hide" forState:UIControlStateNormal];
            
        }
        
    }
}

- (IBAction)showDebugger:(id)sender {
    [[Tapdaq sharedSession] presentDebugViewController];
}

- (IBAction)loadInterstitial:(id)sender {
    [[Tapdaq sharedSession] loadInterstitialForPlacementTag:TDPTagDefault delegate:self];
}

- (IBAction)showInterstitial:(id)sender {
    [self.interstitialAdRequest display];
}

- (IBAction)loadVideo:(id)sender {
    [[Tapdaq sharedSession] loadVideoForPlacementTag:TDPTagDefault delegate:self];
}

- (IBAction)showVideo:(id)sender {
    [self.videoAdRequest display];
}

- (IBAction)loadRewarded:(id)sender {
    [[Tapdaq sharedSession] loadRewardedVideoForPlacementTag:TDPTagDefault delegate:self];
}

- (IBAction)showRewarded:(id)sender {
    [self.rewardedAdRequest display];
}

#pragma mark - Private methods

- (void)logMessage:(NSString *)message
{
    NSString *text = [self.textView text];
    NSString *textAddition = [NSString stringWithFormat:@"%@\n\n%@", message, text];
    [self.textView setText:textAddition];
    
    NSLog(@"Mediation App: %@", message);
}

- (void)updateRewardUI
{
    NSString *rewardName = [[NSUserDefaults standardUserDefaults] objectForKey:kRewardNameKey];
    NSNumber *rewardAmount = [[NSUserDefaults standardUserDefaults] objectForKey:kRewardAmountKey];
    
    if (!rewardName) {
        rewardName = @"Reward";
    }
    
    if (!rewardAmount) {
        rewardAmount = @(0);
    }
    
    [self.rewardName setText:rewardName];
    [self.rewardAmount setText:[rewardAmount stringValue]];
}

#pragma mark - TapdaqDelegate

- (void)didLoadConfig{
    [self logMessage:@"Tapdaq config loaded"];
    [self updateUI];
}

- (void)didFailToLoadConfigWithError:(TDError *)error {
    [self logMessage:@"Tapdaq config failed to load"];
    [self updateUI];
}

#pragma mark - TDAdRequestDelegate

- (void)didLoadAdRequest:(TDAdRequest *)adRequest {
    [self logMessage:[NSString stringWithFormat:@"Did load \"%@\" - tag: \"%@\"", AdTypeNameFromTDAdTypes(adRequest.placement.adTypes), adRequest.placement.tag]];
    if ([adRequest isKindOfClass:[TDMediationAdRequest class]] && [adRequest.placement.tag isEqualToString:TDPTagDefault]) {
        switch (adRequest.placement.adTypes) {
            case TDAdTypeInterstitial:
                self.interstitialAdRequest = (TDMediationAdRequest *)adRequest;
                break;
            case TDAdTypeVideo:
                self.videoAdRequest = (TDMediationAdRequest *)adRequest;
                break;
            case TDAdTypeRewardedVideo:
                self.rewardedAdRequest = (TDMediationAdRequest *)adRequest;
            default:
                break;
        }
    }
    [self updateUI];
}

- (void)adRequest:(TDAdRequest *)adRequest didFailToLoadWithError:(TDError *)error {
    [self logMessage:[NSString stringWithFormat:@"Failed to load \"%@\" - tag: \"%@\"", AdTypeNameFromTDAdTypes(adRequest.placement.adTypes), adRequest.placement.tag]];
    [self logMessage:[NSString stringWithFormat:@"Error: %@", error.localizedDescription]];
}

#pragma mark - TDDisplayableAdRequestDelegate

- (void)willDisplayAdRequest:(TDAdRequest *)adRequest {
    [self logMessage:[NSString stringWithFormat:@"Will display \"%@\" - tag: \"%@\"", AdTypeNameFromTDAdTypes(adRequest.placement.adTypes), adRequest.placement.tag]];
}

- (void)didDisplayAdRequest:(TDAdRequest *)adRequest {
    [self logMessage:[NSString stringWithFormat:@"Did display \"%@\" - tag: \"%@\"", AdTypeNameFromTDAdTypes(adRequest.placement.adTypes), adRequest.placement.tag]];
}

- (void)didCloseAdRequest:(TDAdRequest *)adRequest {
    [self logMessage:[NSString stringWithFormat:@"Did Close \"%@\" - tag: \"%@\"", AdTypeNameFromTDAdTypes(adRequest.placement.adTypes), adRequest.placement.tag]];
}

#pragma mark - TDClickableAdRequestDelegate

- (void)didClickAdRequest:(TDAdRequest *)adRequest {
    [self logMessage:[NSString stringWithFormat:@"Did Click \"%@\" - tag: \"%@\"", AdTypeNameFromTDAdTypes(adRequest.placement.adTypes), adRequest.placement.tag]];
}

#pragma mark - TDRewardedVideoAdRequestDelegate

- (void)adRequest:(TDAdRequest *)adRequest didValidateReward:(TDReward *)reward {
    NSNumber *storedRewardAmount = [[NSUserDefaults standardUserDefaults] objectForKey:kRewardAmountKey];
    
    int storedRewardAmountInt = [storedRewardAmount intValue];
    int updatedRewardAmountInt = storedRewardAmountInt + reward.value;
    
    [[NSUserDefaults standardUserDefaults] setObject:@(updatedRewardAmountInt) forKey:kRewardAmountKey];
    [[NSUserDefaults standardUserDefaults] setObject:reward.name forKey:kRewardNameKey];
    
    if (reward.customJson) {
        [[NSUserDefaults standardUserDefaults] setObject:reward.customJson forKey:kRewardPayloadKey];
    }
    
    [self updateRewardUI];
    
    [self logMessage:[NSString stringWithFormat:@"Received reward for tag:%@ name:%@ value:%d custom JSON: %@", adRequest.placement.tag, reward.name, reward.value, reward.customJson]];
}

- (void)didFailToValidateRewardAdRequest:(TDAdRequest *)adRequest {
    [self logMessage:[NSString stringWithFormat:@"Failed to validate reward for tag:%@", adRequest.placement.tag]];

}
#pragma mark Banner delegate methods

- (void)didLoadBanner
{
    [self updateUI];
    [self logMessage:@"Did load banner"];
}

- (void)didFailToLoadBanner
{
    [self logMessage:@"Failed to load banner"];
}

- (void)didClickBanner
{
    [self logMessage:@"Did click banner"];
}

- (void)didRefreshBanner
{
    [self logMessage:@"Did refresh banner"];
}

@end

File: TapdaqExamples/MediationSwift/ViewController.swift
//
//  ViewController.swift
//  MediationSwift
//
//  Created by Nick Reffitt on 16/05/2017.
//  Copyright © 2017 Tapdaq. All rights reserved.
//

import UIKit
import Tapdaq

func AdTypeNameFromTDAdTypes(types: TDAdTypes) -> String {
    switch types {
    case .typeInterstitial:
        return "interstitial"
    case .typeVideo:
        return "video interstitial"
    case .typeRewardedVideo:
        return "rewarded video"
    default:
        return "unknown"
    }
}

class ViewController: UIViewController, TapdaqDelegate {
    
    struct RewardConstants {
        static let rewardNameKey = "MyRewardName"
        static let rewardAmountKey = "MyRewardAmount"
        static let rewardPayloadKey = "MyRewardPayload"
    }
    
    @IBOutlet var bannerView: UIView!
    var adBanner: UIView?
    
    @IBOutlet var loadInterstitialBtn: UIButton!
    @IBOutlet var loadVideoBtn: UIButton!
    @IBOutlet var loadRewardedBtn: UIButton!
    @IBOutlet var loadBannerBtn: UIButton!
    
    @IBOutlet var showInterstitialBtn: UIButton!
    @IBOutlet var showVideoBtn: UIButton!
    @IBOutlet var showRewardedBtn: UIButton!
    @IBOutlet var showBannerBtn: UIButton!
    
    @IBOutlet var textView: UITextView!
    @IBOutlet var rewardNameLabel: UILabel!
    @IBOutlet var rewardAmountLabel: UILabel!
    
    fileprivate var interstitialAdRequest: TDAdRequest?
    fileprivate var videoAdRequest: TDAdRequest?
    fileprivate var rewardedAdRequest: TDAdRequest?
    

    override func viewDidLoad() {
        super.viewDidLoad()
        Tapdaq.sharedSession().delegate = self
        updateRewardUI()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    fileprivate func updateUI() {
        self.loadInterstitialBtn.isEnabled = Tapdaq.sharedSession()?.isInitialised() ?? false;
        self.loadVideoBtn.isEnabled = Tapdaq.sharedSession()?.isInitialised() ?? false;
        self.loadRewardedBtn.isEnabled = Tapdaq.sharedSession()?.isInitialised() ?? false;
        self.loadBannerBtn.isEnabled = Tapdaq.sharedSession()?.isInitialised() ?? false;
        
        self.showInterstitialBtn.isEnabled = self.interstitialAdRequest?.isReady() ?? false;
        self.showVideoBtn.isEnabled = self.videoAdRequest?.isReady() ?? false;
        self.showRewardedBtn.isEnabled = self.rewardedAdRequest?.isReady() ?? false;
        
        self.showBannerBtn.isEnabled = self.adBanner != nil;
    }
    
    fileprivate func log(message: String) {
        
        let text: String = textView!.text
        let textAddition: String = message + "\n\n" + text
        textView!.text = textAddition
        
        print("Mediation App: " + message)
    }
    
    fileprivate func updateRewardUI() {
 
        let rewardName: String = UserDefaults.standard.string(forKey: RewardConstants.rewardNameKey) ?? "Reward"
        let rewardAmount: Int = UserDefaults.standard.integer(forKey: RewardConstants.rewardAmountKey)

        rewardNameLabel.text = rewardName
        rewardAmountLabel.text = String(rewardAmount)
        
    }
 
    // Target actions
 
    @IBAction func loadInterstitial(sender: UIButton) {
        Tapdaq.sharedSession().loadInterstitial(forPlacementTag: TDPTagDefault, delegate: self)
    }
    
    @IBAction func showInterstitial(sender: UIButton) {
        Tapdaq.sharedSession()?.showInterstitial(forPlacementTag: TDPTagDefault)
    }
    
    @IBAction func loadVideo(sender: UIButton) {
        Tapdaq.sharedSession().loadVideo(forPlacementTag: TDPTagDefault, delegate: self)
    }
    
    @IBAction func showVideo(sender: UIButton) {
        Tapdaq.sharedSession()?.showVideo(forPlacementTag: TDPTagDefault)
    }
    
    @IBAction func loadRewarded(sender: UIButton) {
        Tapdaq.sharedSession().loadRewardedVideo(forPlacementTag: TDPTagDefault, delegate: self)
    }

    @IBAction func showRewarded(sender: UIButton) {
        Tapdaq.sharedSession()?.showRewardedVideo(forPlacementTag: TDPTagDefault)
    }
    
    @IBAction func loadBanner(sender: UIButton) {
        Tapdaq.sharedSession()?.loadBanner(with: .standard, completion: { (bannerView) in
            self.adBanner = bannerView
        })
    }
    
    @IBAction func showBanner(sender: UIButton) {
        
        let tagInt: Int = showBannerBtn.tag
        
        if tagInt == 1 {
            
            adBanner?.removeFromSuperview()
            
            showBannerBtn.tag = 0
            showBannerBtn.setTitle("Show", for: UIControl.State.normal)
            showBannerBtn.isEnabled = false
            
        } else {
            
            let isBannerReady = self.adBanner != nil
            
            if isBannerReady {
                
                bannerView!.addSubview(adBanner!)
                bannerView.setNeedsDisplay()
                
                showBannerBtn.tag = 1
                showBannerBtn.setTitle("Hide", for: UIControl.State.normal)
                
            }
            
        }
        
    }
    
    @IBAction func showDebugger(_ sender: Any?) {
        Tapdaq.sharedSession().presentDebugViewController()
    }
    
    // TapdaqDelegate
    
    func didLoadConfig() {
        
        log(message: "Tapdaq config loaded")
        updateUI()
    }
    
    func didFailToLoadConfigWithError(_ error: TDError!) {
        log(message: "Tapdaq failed to load config with error: \(error?.localizedDescription ?? "")")
        updateUI()
    }
    
    // Banner
    
    func didLoadBanner() {
        updateUI()
        log(message: "Did load banner")
    }
    
    func didFailToLoadBanner() {
        log(message: "Failed to load banner")
    }
    
    func didClickBanner() {
        log(message: "Did click banner")
    }
    
    func didRefreshBanner() {
        log(message: "Did refresh banner")
    }
}

extension ViewController: TDAdRequestDelegate {
    func didLoad(_ adRequest: TDAdRequest) {
        log(message: "Did load \"\(AdTypeNameFromTDAdTypes(types: adRequest.placement.adTypes))\" - tag: \"\(adRequest.placement.tag ?? "")\"")
        switch (adRequest.placement.adTypes) {
        case .typeInterstitial:
            interstitialAdRequest = adRequest
        case .typeVideo:
            videoAdRequest = adRequest
        case .typeRewardedVideo:
            rewardedAdRequest = adRequest
        default:
            break;
        }
        
        updateUI()
    }
    
    func adRequest(_ adRequest: TDAdRequest, didFailToLoadWithError error: TDError?) {
        log(message: "Failed to load \"\(AdTypeNameFromTDAdTypes(types: adRequest.placement.adTypes))\" - tag: \"\(adRequest.placement.tag ?? "")\" error: \(error?.localizedDescription ?? "")")
        updateUI()
    }
}

extension ViewController: TDDisplayableAdRequestDelegate {
    func willDisplay(_ adRequest: TDAdRequest) {
        log(message: "Will display\"\(AdTypeNameFromTDAdTypes(types: adRequest.placement.adTypes))\" - tag: \"\(adRequest.placement.tag ?? "")\"")
    }
    
    func didDisplay(_ adRequest: TDAdRequest) {
        log(message: "Did display\"\(AdTypeNameFromTDAdTypes(types: adRequest.placement.adTypes))\" - tag: \"\(adRequest.placement.tag ?? "")\"")
    }
    
    func adRequest(_ adRequest: TDAdRequest, didFailToDisplayWithError error: TDError?) {
        log(message: "Failed to display \"\(AdTypeNameFromTDAdTypes(types: adRequest.placement.adTypes))\" - tag: \"\(adRequest.placement.tag ?? "")\" error: \(error?.localizedDescription ?? "")")
        updateUI()
    }
    
    func didClose(_ adRequest: TDAdRequest) {
        log(message: "Did close \"\(AdTypeNameFromTDAdTypes(types: adRequest.placement.adTypes))\" - tag: \"\(adRequest.placement.tag ?? "")\"")
    }
}

extension ViewController: TDClickableAdRequestDelegate {
    func didClick(_ adRequest: TDAdRequest) {
        log(message: "Did click \"\(AdTypeNameFromTDAdTypes(types: adRequest.placement.adTypes))\" - tag: \"\(adRequest.placement.tag ?? "")\"")
    }
}

extension ViewController: TDRewardedVideoAdRequestDelegate {
    func adRequest(_ adRequest: TDAdRequest, didValidate reward: TDReward) {
        let storedRewardAmount: Int = UserDefaults.standard.integer(forKey: RewardConstants.rewardAmountKey)
        
        let updatedRewardAmount: Int = storedRewardAmount + Int(reward.value)
        
        UserDefaults.standard.set(updatedRewardAmount, forKey: RewardConstants.rewardAmountKey)
        UserDefaults.standard.set(reward.name, forKey: RewardConstants.rewardNameKey)
        if let _ = reward.customJson {
            UserDefaults.standard.set(reward.customJson, forKey: RewardConstants.rewardPayloadKey)
        }
        updateRewardUI()
        
        log(message: "Received reward for tag:\(adRequest.placement.tag ?? "") name:\(reward.name) value:\(reward.value) custom JSON: \(reward.customJson)")
    }
    
    func didFail(toValidateRewardAdRequest adRequest: TDAdRequest) {
        log(message: "Failed to validate reward for tag: \(adRequest.placement.tag ?? "")")
    }
}

extension ViewController: TDOfferwallAdRequestDelegate {
    func adRequest(_ adRequest: TDAdRequest, didReceiveOfferwallCredits creditInfo: [AnyHashable : Any]?) {
        log(message: "Received offerwall credits: \(creditInfo ?? [:])")
    }
}
